### ISSUE TITLE
Support Devfile format to generate a workspace

### ISSUE THREAD

**Issue by slemeur** [eclipse-che/che#11549]
# Goal
Today we have two kind of recipes to bootstrap a cloud developer workspace and make it portable: [Chefile](https://www.eclipse.org/che/docs/chefile.html) and [Factories](https://www.eclipse.org/che/docs/factories-getting-started.html#try-a-factory).

Over the last few month, we've been also iterating on the definition of a `devfile` format which defines the different tools needed to develop a container based application. That work has been initiated along with the requirements of [Workspace.Next](https://github.com/eclipse/che/issues/8265). You can find the defined format here: https://github.com/l0rd/devfile/

The goal of this epic is to support the `devfile` format in Eclipse Che on all the different ways we have to create/generate a workspace. 

## User Stories
With the following requirements:
- a repository like `github.com/che-samples/spring-petclinic`
- a `devfile` hosted under the repository

**User Stories:**
1. A developer can generate a workspace by just calling a url: `che.host/f?repo_url`
2. A developer can create a workspace by providing the url to the repository
3. A developer can create a workspace by importing the `devfile`
4. A developer can generate a workspace by sending a devfile to a rest API

**Not in scope for now:**
1. A developer can generate the `devfile` based on an existing workspace
2. A developer can import a repository without a `devfile` and be guided to select the tools he needs to create the properly tooled workspace
3. A developer can change the configuration of the workspace and changes are reflected + versioned in the devfile
4. A developer should be exposed to the devfile format when he is configuring a workspace
5. A developer can create a workspace based on multi-repo with multi `devfiles`

## Sub-Tasks

**A developer can generate a workspace by just calling a url: `che.host/f?repo_url`**
A developer should be able to get a workspace from a repository where there is a `devfile`. Today a user is able to get a workspace by giving the url of the repository to the factory engine. It should be able to understand if the repository is having a `devfile` and if it has, use it to create the properly tooled workspace.

**A developer can create a workspace by providing the url to the repository**
We have different ways to create a new workspace from the dashboard. The user should be able to create a workspace by providing the URL to a repository. Che will be able to create the workspace from a project's `devfile` hosted in the repository. 

- In the workspace creation page, when user provides the URL to a repository with a `devfile`, the `devfile` should be used to configure the workspaces.

**A developer can create a workspace by importing the `devfile`**
A developer should be able to provide the content of a `devfile` or import a `devfile` to create a workspace.  

- In the workspace list, we could add an "import" button which will allow to import a `devfile` either by uploading or copy/pasting the content of the file.

**A developer can generate a workspace by sending a devfile to a rest API**
A developer should be able to send to a restAPI:
- a devfile
- a url of a repository with a devfile
An get as a result a workspace created and the url to access it.

## Phase 1
- [x] Setup Devfile implementation plan fro Phase 1 https://github.com/eclipse/che/issues/11719 
- [x] Allow running Che 7 Workspace with development tooling but without recipe https://github.com/eclipse/che/issues/11690
- [x] Commands API for Workspace.Next https://github.com/eclipse/che/issues/9546
- [x] Implementation of local part of devfile feature  https://github.com/eclipse/che/issues/11832
- [x] Create a JSON-schema for the Devfile  https://github.com/eclipse/che/issues/11904

## Phase 2
- [ ] command matches against a tool, but the tool can have multiple containers
- [x] Devfile support in factory https://github.com/eclipse/che/issues/12186
- [x] Plan and POC implementation of remote part of devfile feature https://github.com/eclipse/che/issues/11833 
- [x] An ability to specify workspace environment in Devfile https://github.com/eclipse/che/issues/11906
- [x] IDE should respect commands from Runtime https://github.com/eclipse/che/issues/12053
- [x] Devfile Spec documentation generator https://github.com/eclipse/che/issues/12272
- [x] Add ability to specify workspace environment in Devfile #11906 ([doc](https://docs.google.com/document/d/1ywDiI4ZWq81w32d3bIkYe1PY1m_ANxSLXBzRHi5MYsc/edit?usp=sharing), [PR1](https://github.com/redhat-developer/devfile/pull/2))
   - [x] An ability to specify Kubernetes/OpenShift List as tool in Devfile https://github.com/eclipse/che/issues/11906
   - [x] Factory flow should support k8s/OS tool in Devfile https://github.com/eclipse/che/issues/12266
   - [x] K8s/OS infrastructure should use PVCs as Che Volumes https://github.com/eclipse/che/issues/12269
   - [x] K8s/OS infrastructure should use Ingresses/Service/Route as Che Server https://github.com/eclipse/che/issues/12268 
- [x] [Devfile] Support commands for Os/K8s tool https://github.com/eclipse/che/issues/12479
- [x] Implement a way for the devfile user to specify the root path of the projects https://github.com/eclipse/che/issues/12531
- [x] Publish Devfile documentation on http://redhat-developer.github.io/devfile https://github.com/eclipse/che/issues/12524 
- [x] Support mountSources for non-plugin tools in devfile https://github.com/eclipse/che/issues/12554
- [x] Support custom environment variables for tools in devfile https://github.com/eclipse/che/issues/12555
- [x] Support checkoutPath in devfile  https://github.com/eclipse/che/issues/12557
- [x] Support autodetection of checkoutPath of projects  https://github.com/eclipse/che/issues/12558
- [x] REST API to create workspaces from a devfile https://github.com/eclipse/che/issues/12572
- [x] Study/Devfile: proper object hierarchy for tools in the schema and java model  https://github.com/eclipse/che/issues/12587
- [x] Add support of dockerimage tool in Devfile https://github.com/eclipse/che/issues/12389
- [x] Add an ability to specify entrypoint for dockerimage tool in Devfile https://github.com/eclipse/che/issues/12668
- [x] Implement an ability to specify container for commands for K8s/OS tool with multiple containers https://github.com/eclipse/che/issues/12606
- [x] Make it possible to specify path for command execution in its configuration https://github.com/eclipse/che/issues/11866
- [x] Support workspace creating from repository if it contains `.devfile.yaml`/`devfile.yaml` file https://github.com/eclipse/che/issues/12675
- [x] Support devfile with more than one Kubernetes tool https://github.com/eclipse/che/issues/12719
- [x] Devfile on-boarding and learning experience  https://github.com/eclipse/che/issues/12732 
- [x] Add an ability to specify entrypoint for kube and openshift tool in Devfile https://github.com/eclipse/che/issues/12693
- [x] [Devfile] OpenShift/K8s tool - reference to content  https://github.com/eclipse/che/issues/12733
- [x] Inform users about what devifle was used to create a workspace with the factory | https://github.com/eclipse/che/issues/12741
- [x] [devfile] Che commands (Theia task) targeting a container tool defined by url is not working https://github.com/eclipse/che/issues/12765
 - [x] Add more flexibility into devfile project source - checkout branch, commit ID etc 
https://github.com/eclipse/che/issues/12775
- [x] Support resolving referenced files when building a factory from a link to raw content of the devfile 
 https://github.com/eclipse/che/issues/12798
- [x] Default devfile name https://github.com/eclipse/che/issues/12837
- [x] Devfile: tools should be renamed components https://github.com/eclipse/che/issues/12836
- [x] Make it possible to add exposed ports for container in Devfile https://github.com/eclipse/che/issues/12808
- [x] Support multiple dockerimage tools in Devfile https://github.com/eclipse/che/issues/12875

**Comment by slemeur** [comment]
Related to #4362

**Comment by l0rd** [comment]
Some thoughts:

- I think that stories 1. and 4. are related (it's about a new API), and 2. and 3. are related as well (about the client that call these API, the user dashboard). I think that 1. and 4. are prerequisites and should be handled with higher priority. And probably 2. and 3. will be implemented by a different team. Maybe we should split them in 2 different epics? 

- What it's expected if a developer modify the devfile from Che IDE? Or if he creates a new devfile? Should the workspace config be changed and the user asked if he wants to reload the workspace? If not what will happen when the workspace is stopped and restarted? What if the URL of the github repo in the devfile (the project source) is changed?

- It looks like there isn't nothing to change with respect to chefile and factories. Devfile will be an alternative to these two features. Is it correct? Maybe we could make it explicit

**Comment by skabashnyuk** [comment]
Before we start use-case grouping I would like to have an agreement on some technical aspects.

I would say all stories can be divided into such categories.

- we have Devfile content
- we have an HTTP link to Devfile
- Devfile is under vcs. (This use case has the biggest effect on technical solution)

Next, we have two ways to have a working workspace from Devfile.
1. Get content of Devfile, transform it to workspace config 1 to 1 before workspace start.
2. Get content of Devfile during workspace start

From what I see, I think the second solution is better, especially for the most interesting case when Devfile is under vcs.


My proposal is to add new optional entry to workspace configuration.

1. For the case when we have Devfile content
```
        "devfile": {
          "type": "devfile",
          "content": "Content of devfile"
          "parameters": {}
        }
```


2. For the case when we have a http link for Devfile
```
        "devfile": {
          "type": "url",
          "content": "https://github.com/che-samples/console-java-simple/Devfile"
          "parameters": {}
        }
```

3. For the case when Devfile is under vcs
```
        "devfile": {
          "type": "git",
          "content": "https://github.com/che-samples/console-java-simple.git"
          "parameters": {}
        }
```
```
        "devfile": {
          "type": "hg",
          "content": "http://selenic.com/hg"
          "parameters": {}
        }
```
```
        "devfile": {
          "type": "svn",
          "content": "svn://xxx.xxx.xxx.xxx/Repository/"
          "parameters": {}
        }
```
Then on workspace start, we will have some DevFile Broker(dependent on type) that will play alongside with our ChePlugin broker to get effective workspace configuration. 
The questions here:

- who, when and how many times are going to clone sources.
- how to generate Devfile from workspace with Devfile under vcs.

**Comment by sleshchenko** [comment]
@skabashnyuk @slemeur @l0rd I believe it's a good time to revise this epic and close it, maybe create another one for some improvements since every issue linked here is fixed or closed by che-bot because of inactivity for 180 days.
